### PR TITLE
[trajoptlib] Use thiserror for error types

### PR DESCRIPTION
**Pull request by spacey-sooty**

### CODE REVIEW

**Comment by oh-yes-0-fps** [comment]
What is this waiting on or what still needs to be done?

**Comment by spacey-sooty** [comment]
Fixing conflicts, properly propogating the error from C++ -> Rust and API design decision. I'm gonna take another look at this in a bit to fix some of this stuff up

File: trajoptlib/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+use thiserror::Error;
+
+#[derive(Debug, Error)]
+pub enum TrajoptError {
+    #[error("The solver determined the problem to be overconstrained and gave up.")]
```
Comment (calcmogul) [inline trajoptlib/src/error.rs:5]: The fact this duplicates the contents of https://github.com/SleipnirGroup/Sleipnir/blob/main/include/sleipnir/optimization/SolverExitCondition.hpp#L47-L78 isn't ideal.

Comment (spacey-sooty) [reply trajoptlib/src/error.rs:5]: yeah, do you have a better option?

Comment (calcmogul) [reply trajoptlib/src/error.rs:5]: See my edited message.

**Review by oh-yes-0-fps — APPROVED** [review]

**Pull request merged**